packages/svelte/src/internal/client/proxy.js

```diff
@@ -24,273 +22,268 @@ import * as e from './errors.js';
  * @returns {ProxyStateObject<T> | T}
  */
 export function proxy(value, parent = null, prev) {
-	if (typeof value === 'object' && value != null && !is_frozen(value)) {
-		// If we have an existing proxy, return it...
-		if (STATE_SYMBOL in value) {
-			const metadata = /** @type {ProxyMetadata<T>} */ (value[STATE_SYMBOL]);
-
-			// ...unless the proxy belonged to a different object, because
-			// someone copied the state symbol using `Reflect.ownKeys(...)`
-			if (metadata.t === value || metadata.p === value) {
-				if (DEV) {
-					// Since original parent relationship gets lost, we need to copy over ancestor owners
-					// into current metadata. The object might still exist on both, so we need to widen it.
-					widen_ownership(metadata, metadata);
-					metadata.parent = parent;
-				}
+	// if non-proxyable, or is already a proxy, return `value`
+	if (typeof value !== 'object' || value === null || STATE_SYMBOL in value) {
+		return value;
+	}
 
-				return metadata.p;
-			}
+	const prototype = get_prototype_of(value);
+
+	if (prototype !== object_prototype && prototype !== array_prototype) {
+		return value;
+	}
+
+	var sources = new Map();
+	var is_proxied_array = is_array(value);
+	var version = source(0);
+
+	/** @type {ProxyMetadata} */
+	var metadata;
+
+	if (DEV) {
+		metadata = {
+			parent,
+			owners: null
+		};
+
+		if (prev) {
+			// Reuse owners from previous state; necessary because reassignment is not guaranteed to have correct component context.
+			// If no previous proxy exists we play it safe and assume ownerless state
+			// @ts-expect-error
+			const prev_owners = prev.v?.[STATE_SYMBOL_METADATA]?.owners;
+			metadata.owners = prev_owners ? new Set(prev_owners) : null;
+		} else {
+			metadata.owners =
+				parent === null
+					? current_component_context !== null
+						? new Set([current_component_context.function])
+						: null
+					: new Set();
 		}
+	}
 
-		const prototype = get_prototype_of(value);
-
-		if (prototype === object_prototype || prototype === array_prototype) {
-			const proxy = new Proxy(value, state_proxy_handler);
-
-			define_property(value, STATE_SYMBOL, {
-				value: /** @type {ProxyMetadata} */ ({
-					s: new Map(),
-					v: source(0),
-					a: is_array(value),
-					p: proxy,
-					t: value
-				}),
-				writable: true,
-				enumerable: false
-			});
+	return new Proxy(/** @type {any} */ (value), {
+		defineProperty(_, prop, descriptor) {
+			if (
+				!('value' in descriptor) ||
+				descriptor.configurable === false ||
+				descriptor.enumerable === false ||
+				descriptor.writable === false
+			) {
+				e.state_descriptors_fixed();
```

Comment:
would be good to have a comment here explaining why we have this restriction

Reply:
Thanks for the comment - reading up on the invariants, can we relax/change this validation? IIUC we only need to disallow non-configurable properties, because they need to appear as such on the target, too. On the other hand, we cannot allow it at all if the target is no extensible and doesn't contain this property

Reply:
The invariant is correct. We also can’t support non enumerable properties or non-writable because then we need extra overhead to store the extra descriptor info per lookup. See the pr that this came from for context.

Reply:
Fair enough - for future reference, this is where it original came up: https://github.com/sveltejs/svelte/pull/12847#issuecomment-2290432963

Reply:
Thanks for that, was on mobile earlier so couldn't link it.